server/service/src/report/report_service.rs

```diff
@@ -96,74 +96,126 @@ pub trait ReportServiceTrait: Sync + Send {
-            .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
-        let now: DateTime<Utc> = SystemTime::now().into();
-        let file = file_service
-            .store_file(
-                &format!("{}_{}.pdf", now.format("%Y%m%d_%H%M%S"), report_name),
-                &pdf,
-            )
-            .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
-        Ok(file.id)
-    }
+/// Converts a HTML report to a pdf file and returns the file id
+fn print_html_report_to_pdf(
+    base_dir: &Option<String>,
+    document: GeneratedReport,
+    report_name: String,
+) -> Result<String, ReportError> {
+    let id = uuid();
+    // TODO use a proper tmp dir here instead of base_dir?
+    let pdf = html_to_pdf(base_dir, &format_html_document(document), &id)
+        .map_err(|err| ReportError::HTMLToPDFError(format!("{}", err)))?;
+
+    let file_service = StaticFileService::new(base_dir)
+        .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
+    let now: DateTime<Utc> = SystemTime::now().into();
+    let file = file_service
+        .store_file(
+            &format!("{}_{}.pdf", now.format("%Y%m%d_%H%M%S"), report_name),
+            &pdf,
+        )
+        .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
+    Ok(file.id)
+}
 
-    fn print_html_report(
-        &self,
-        base_dir: &Option<String>,
-        document: GeneratedReport,
-        report_name: String,
-    ) -> Result<String, ReportError> {
-        let file_service = StaticFileService::new(base_dir)
-            .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
-        let now: DateTime<Utc> = SystemTime::now().into();
-        let file = file_service
-            .store_file(
-                &format!("{}_{}.html", now.format("%Y%m%d_%H%M%S"), report_name),
-                document.document.as_bytes(),
-            )
-            .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
-        Ok(file.id)
-    }
+/// Converts the report to a HTML file and returns the file id
+fn print_html_report_to_html(
+    base_dir: &Option<String>,
+    document: GeneratedReport,
+    report_name: String,
+) -> Result<String, ReportError> {
+    let file_service = StaticFileService::new(base_dir)
+        .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
+    let now: DateTime<Utc> = SystemTime::now().into();
+    let file = file_service
+        .store_file(
+            &format!("{}_{}.html", now.format("%Y%m%d_%H%M%S"), report_name),
+            format_html_document(document).as_bytes(),
+        )
+        .map_err(|err| ReportError::DocGenerationError(format!("{}", err)))?;
+    Ok(file.id)
+}
+
+/// Puts the document content, header and footer into a <html> template.
+/// Its assumed
+fn format_html_document(document: GeneratedReport) -> String {
+    format!(
+        "<html>
+<head>
+<style>
```

Comment:
can you use the suggestion from the previous review please?
I don't believe the `1.5cm` margins are needed - for my testing, the height was adjusting correctly and the headers & footers repeated

```
<html>
  <body>
      <table class=\"paging\">
      <thead>
        <tr>
          <td>{}</td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>{}</td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td>{}</td>
        </tr>
      </tfoot>
    </table>
  </body>
</html>
```

Reply:
Ok thank I missed this earlier. I used your code now.